### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by shivanshuag**
Issue #42 
Instruction to run the tests
- Run `composer update` inside site_audit directory to install all the dependencies.
- Run `composer dump-autoload` to regnerate the autoload file.
- Export the environment variable `UNISH_DB_URL=mysql://[username]:[password]@HOST` with the correct mysql username, password and host for PHPUnit to use for testing.
- Run the script `./test.sh` from inside the `site_audit` directory. This will run all the tests.
- To run the tests selectively, run `./test.sh --filter="*Extensions*" tests` where `*Extensions*` is the regex matching the class name of tests for the extension report.

### CODE REVIEW

**Comment by shivanshuag** [comment]
For setting the config varibles, the `drush_config_set` function cannot be used from inside the tests. The only other option is using the `drush config-set` command which does not provide any improvement in the time taken by the tests to run.

File: tests/bestPracticesReportTest.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  /**
+   * If fast_404 is enabled and fast_404 paths are empty, check should warn.
+   */
+  public function testFast404One() {
+    // Enable fast_404 and make fast_404 paths empty.
+    $eval1 = <<<EOT
+\$config = \\Drupal::configFactory()->getEditable('system.performance');
+\$config->set('fast_404.enabled', TRUE);
+\$config->save();
+EOT;
```
Comment (fluxsauce) [inline tests/bestPracticesReportTest.php]: No need to open and close.

File: tests/bestPracticesReportTest.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+  /**
+   * Fast_404 enabled and fast_404 paths not empty in fast_404, check passes.
+   */
+  public function testFast404Two() {
+    $eval1 = <<<EOT
+\$config = \\Drupal::configFactory()->getEditable('system.performance');
+\$config->set('fast_404.enabled', TRUE);
+\$config->save();
+EOT;
```
Comment (fluxsauce) [inline tests/bestPracticesReportTest.php]: No need to open and close.

**Comment by fluxsauce** [comment]
``` bash
jon@Decker ~/projects/site_audit (unit-tests) $ ./test.sh
./test.sh: line 32: phpunit: command not found
jon@Decker ~/projects/site_audit (unit-tests) $ ls -la vendor/bin/
total 24
drwxr-xr-x   5 jon  staff  170 Jul 27 09:49 .
drwxr-xr-x  13 jon  staff  442 Jul 27 09:49 ..
lrwxr-xr-x   1 jon  staff   43 Jul  7 10:46 phpcbf -> ../squizlabs/php_codesniffer/scripts/phpcbf
lrwxr-xr-x   1 jon  staff   42 Jul  7 10:46 phpcs -> ../squizlabs/php_codesniffer/scripts/phpcs
lrwxr-xr-x   1 jon  staff   26 Jul 27 09:49 phpunit -> ../phpunit/phpunit/phpunit
```

Similar to `phpcs.sh`, check to see if `phpunit` is installed in `/vendor/bin` and abort if it's not. Only use the version of `phpunit` that is installed with Composer in the local directory.

**Comment by fluxsauce** [comment]
Trying a workaround and it's not working.

```
if [ $# = 0 ] ; then
   ./vendor/bin/phpunit --configuration="$DRUSH_DIRNAME/tests" tests
else
   # Pass along any arguments.
   ./vendor/bin/phpunit --configuration="$DRUSH_DIRNAME/tests" $@
fi
```

resulted in

```
jon@Decker ~/projects/site_audit (unit-tests) $ ./test.sh
PHP Fatal error:  Class 'Unish\UnishTestCase' not found in /Users/jon/.composer/vendor/drush/drush/tests/bootstrap.inc on line 51

Fatal error: Class 'Unish\UnishTestCase' not found in /Users/jon/.composer/vendor/drush/drush/tests/bootstrap.inc on line 51
```

**Comment by shivanshuag** [comment]
> Trying a workaround and it's not working.

This works for me. Maybe there is some difference in our setup. Did you run `composer install` inside site_audit directory?

**Comment by fluxsauce** [comment]
> This works for me. Maybe there is some difference in our setup. Did you run composer install inside site_audit directory?

Composer update.

``` bash
jon@Decker ~/projects/site_audit (8.x-2.x) $ composer update
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Installing sebastian/version (1.0.6)
    Loading from cache

  - Installing sebastian/global-state (1.0.0)
    Loading from cache

  - Installing sebastian/recursion-context (1.0.1)
    Downloading: 100%

  - Installing sebastian/exporter (1.2.1)
    Downloading: 100%

  - Installing sebastian/environment (1.3.0)
    Downloading: 100%

  - Installing sebastian/diff (1.3.0)
    Loading from cache

  - Installing sebastian/comparator (1.2.0)
    Downloading: 100%

  - Installing symfony/yaml (v2.7.2)
    Loading from cache

  - Installing doctrine/instantiator (1.0.5)
    Loading from cache

  - Installing phpdocumentor/reflection-docblock (2.0.4)
    Loading from cache

  - Installing phpspec/prophecy (v1.4.1)
    Loading from cache

  - Installing phpunit/php-text-template (1.2.1)
    Loading from cache

  - Installing phpunit/phpunit-mock-objects (2.3.6)
    Downloading: 100%

  - Installing phpunit/php-timer (1.0.7)
    Downloading: 100%

  - Installing phpunit/php-token-stream (1.4.3)
    Loading from cache

  - Installing phpunit/php-file-iterator (1.4.1)
    Downloading: 100%

  - Installing phpunit/php-code-coverage (2.1.9)
    Downloading: 100%

  - Installing phpunit/phpunit (4.7.7)
    Downloading: 100%

sebastian/global-state suggests installing ext-uopz (*)
phpdocumentor/reflection-docblock suggests installing dflydev/markdown (~1.0)
phpdocumentor/reflection-docblock suggests installing erusev/parsedown (~1.0)
phpunit/php-code-coverage suggests installing ext-xdebug (>=2.2.1)
phpunit/phpunit suggests installing phpunit/php-invoker (~1.1)
Writing lock file
Generating autoload files
jon@Decker ~/projects/site_audit (unit-tests) $ ./test.sh
./test.sh: line 32: phpunit: command not found
```

Trying install:

``` bash
jon@Decker ~/projects/site_audit (unit-tests) $ composer install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Nothing to install or update
Generating autoload files
jon@Decker ~/projects/site_audit (unit-tests) $ ./test.sh
PHP Fatal error:  Class 'Unish\UnishTestCase' not found in /Users/jon/.composer/vendor/drush/drush/tests/bootstrap.inc on line 51

Fatal error: Class 'Unish\UnishTestCase' not found in /Users/jon/.composer/vendor/drush/drush/tests/bootstrap.inc on line 51
```

**Comment by shivanshuag** [comment]
Sorry. I meant `composer update`. One more thing, run `composer update` inside drush directory also. I am not sure why my setup works. I was getting similar error when I started writing unit tests and did many things after that, one of which made it work. Unfortunately, i don't know which one!

**Comment by fluxsauce** [comment]
> Unfortunately, i don't know which one!

I'm sorry to hear that, but functional testing is a requirement - if I can't run it with the instructions provided, how could anyone else?

**Comment by shivanshuag** [comment]
I investigated, on removing the vendor folder inside the drush directory, I got the same error. So, running `composer update` inside the drush directory should solve the issue.

**Comment by fluxsauce** [comment]
```
jon@Decker ~/projects/site_audit (unit-tests) $ ./test.sh
./test.sh: line 32: phpunit: command not found
jon@Decker ~/projects/site_audit (unit-tests) $ ls -la vendor/bin/
total 24
drwxr-xr-x   5 jon  staff  170 Jul 29 08:58 .
drwxr-xr-x  13 jon  staff  442 Jul 29 08:58 ..
lrwxr-xr-x   1 jon  staff   43 Jul 28 14:05 phpcbf -> ../squizlabs/php_codesniffer/scripts/phpcbf
lrwxr-xr-x   1 jon  staff   42 Jul 28 14:05 phpcs -> ../squizlabs/php_codesniffer/scripts/phpcs
lrwxr-xr-x   1 jon  staff   26 Jul 29 08:58 phpunit -> ../phpunit/phpunit/phpunit
```

**Comment by fluxsauce** [comment]
Getting there!

> Make sure PHPUnit is installed and in the PATH.

That's not the right way to do it, it's installed with Composer. All the requirements should be installed using composer, which includes Drush. I fixed the PHPUnit part, but you should correct the drush part.

As we discussed last week:
- In the bash script, check to see if the commands exist, and if they don't, bail.

Additionally:
- Include drush as a composer dev requirement
- Use composer path to drush (don't try to detect it)
- Your instructions on environmental variables are not explicit; update the README:

``` bash
export UNISH_DB_URL="mysql://USERNAME:PASSWORD@HOST"
```

For some reason, there's a failure when using the stop on failure mode?

``` bash
jon@Decker ~/projects/site_audit (unit-tests) $ ./test.sh
PHPUnit 4.7.7 by Sebastian Bergmann and contributors.

...................................|

Time: 5.64 minutes, Memory: 8.00Mb

OK (35 tests, 35 assertions)
jon@Decker ~/projects/site_audit (unit-tests) $ ./test.sh --stop-on-failure
PHPUnit 4.7.7 by Sebastian Bergmann and contributors.

..........I...........F-

Time: 3.26 minutes, Memory: 32.25Mb

There was 1 failure:

1) Unish\cacheCommandCase::testCacheGet
Unexpected exit code 1 (expected 0) for command:
/Users/jon/.composer/vendor/drush/drush/drush --root=/private/var/folders/8s/f34z4myx2fq9tzk8h08_2mf80000gp/T/drush-sandbox/web --uri=dev --nocolor cache-get system.filter config --yes --format=json

Command stderr:
The system.filter object in the config bin was not found.                [error]

/Users/jon/.composer/vendor/drush/drush/tests/Unish/CommandUnishTestCase.php:190
/Users/jon/.composer/vendor/drush/drush/tests/Unish/CommandUnishTestCase.php:293
/Users/jon/.composer/vendor/drush/drush/tests/cacheCommandTest.php:27
/Users/jon/.composer/vendor/drush/drush/tests/Unish/CommandUnishTestCase.php:322
/Users/jon/projects/site_audit/vendor/phpunit/phpunit/src/TextUI/Command.php:147
/Users/jon/projects/site_audit/vendor/phpunit/phpunit/src/TextUI/Command.php:99

FAILURES!
Tests: 23, Assertions: 51, Failures: 1, Incomplete: 1.
```

**Comment by shivanshuag** [comment]
> For some reason, there's a failure when using the stop on failure mode?

The usage of the script is wrong. When specifying extra arguments to `test.sh` script, you also need to give the directory containing the tests. So running `./test.sh --stop-on-failure` ran the drush tests, not site_audit tests. Please run `./test.sh --stop-on-failure tests`.

**Comment by shivanshuag** [comment]
@fluxsauce Please run `composer dump-autoload` along with `composer update` to test this. I have added an autoload entry for `Unish` namespace inside composer.json

**Pull request merged**